### PR TITLE
feat(ci): CircleCI migration to Github Actions

### PR DESCRIPTION
**Pull request by toninis**
#### Summary

Migrate from CircleCI to Github Actions:

- Migrated nightly builds to schedules on Github . We are building every night from **master** branch so there is no need to create branches and push tags from gitlab any more . This whole pipeline will be deleted . We added a small action to patch the **package.json** for the nightly builds . @devinbinnie for feedback on this approach
- We have not yet implemented the channel notification with the nightly builds since it needs refactoring and some extra API calls on Github in order to generate separate links per build. @devinbinnie feedback on this in order to start implementation since we may have to redirect those links from S3 since we upload them in any case . 

#### Ticket Link

https://mattermost.atlassian.net/browse/CLD-4812

#### Release Note

```release-note
Migrated from CircleCI to Github Actions
```

### CODE REVIEW

**Comment by devinbinnie** [comment]
cc @wget in case any of this affects the work you're doing :)

**Comment by wget** [comment]
@devinbinnie Great, thanks a lot :) I was just reading how to do this with GitHub Actions.
I propose we merge @toninis's work and I merge my changes upon :)

File: .circleci/config.yml
Comment (devinbinnie) [inline .circleci/config.yml:603]: What's this bit for? Just a test or something?

File: .github/actions/patch-nightly-version/patch-version.go
Comment (devinbinnie) [inline .github/actions/patch-nightly-version/patch-version.go:1]: Is there a reason we can't use the original version patching code? It feels weird to be injecting Golang code into a mostly JS/TS based project.

**Review by devinbinnie — COMMENTED** [review]
Looking good so far! One question: are we able to download the builds from the CI build itself like we could on CircleCI?

The scheduling of the nightly build looks good to me.
Let me know what help I can provide for the channel notifications, I thought we just needed to hit a webhook on community.

File: .github/actions/patch-nightly-version/patch-version.go
Comment (toninis) [reply .github/actions/patch-nightly-version/patch-version.go:1]: Is there a specific command/script you are thinking of ? I grabbed this logic from the trigger nightly built from Gitlab repo

File: .circleci/config.yml
Comment (toninis) [reply .circleci/config.yml:603]: Yep . The whole file will be removed after we test that everything is in place

Comment (toninis) [reply .circleci/config.yml:603]: > Looking good so far! One question: are we able to download the builds from the CI build itself like we could on CircleCI?
> 
> The scheduling of the nightly build looks good to me. Let me know what help I can provide for the channel notifications, I thought we just needed to hit a webhook on community.

@devinbinnie Let's sync together on this . There is a difference on how Github handles artifacts vs CirlceCI so I would like your input

**Comment by toninis** [comment]
> Looking good so far! One question: are we able to download the builds from the CI build itself like we could on CircleCI?
> 
> The scheduling of the nightly build looks good to me. Let me know what help I can provide for the channel notifications, I thought we just needed to hit a webhook on community.

@devinbinnie Let's sync together on this . There is a difference on how Github handles artifacts vs CirlceCI so I would like your input

**Comment by toninis** [comment]
@devinbinnie Everything is set up from my side . There are some tests that are failing . 
Can you take a look there ? For some reason `bg.json` file is removed and `i18n.ts` is not updated . 

Thanks 😄

**Review by devinbinnie — APPROVED** [review]
All is looking good to me! Thanks @toninis for the work done on this. Approving so as not to block since we may need to iterate a bit.

One question: I'm still not seeing any builds in the `ci` workflows artifacts output, is that coming later?

**Comment by toninis** [comment]
@devinbinnie Please merge when you are ready and tested everything needed

**Comment by devinbinnie** [comment]
> @devinbinnie Please merge when you are ready and tested everything needed

@toninis Should we not wait for other reviews first?

**Comment by toninis** [comment]
> > @devinbinnie Please merge when you are ready and tested everything needed
> 
> 
> 
> @toninis Should we not wait for other reviews first? 

Yep. My team will also review of course .

File: .github/actions/patch-nightly-version/action.yaml
Comment (phoinixgrr) [inline .github/actions/patch-nightly-version/action.yaml:11]: Could we just run something like the following, and remove all this GoLang clutter? 

```
cat <<< $(jq '.version = (.version | split("-") | first | sub("$"; "-nightly.") | sub("$"; (now | strflocaltime("%Y%m%d"))))' package.json) > package.json
```

```suggestion
      run: cat <<< $(jq '.version = (.version | split("-") | first | sub("$"; "-nightly.") | sub("$"; (now | strflocaltime("%Y%m%d"))))' package.json) > package.json
```

File: .github/workflows/nightly-browser-view.yml
Comment (phoinixgrr) [inline .github/workflows/nightly-browser-view.yml:228]: Building Json under Bash can be error prone. 

Maybe consider replacing this with: 
```suggestion
      jq --null-input \
              --arg icon_url "https://upload.wikimedia.org/wikipedia/commons/1/17/Luna_symbol.png" \
              --arg username "NightBuilder" \
              --arg text "${{ needs.upload-to-s3.outputs.links }}" \
              '{"username":$username,"icon_url": $icon_url, "text": $text }' > /tmp/webhook-data.json
```

File: .github/workflows/nightly-rainforest.yml
Comment (phoinixgrr) [inline .github/workflows/nightly-rainforest.yml:120]: Maybe decouple the version extract with a variable, and use the Variable to inject this to the command. 
Would something like the following work: ? 
```
env:
  APP_VERSION: $(jq -r .version package.json)
```

File: .github/workflows/nightly-rainforest.yml
Comment (phoinixgrr) [inline .github/workflows/nightly-rainforest.yml:9]: This can be removed, since bash is the default shell for linux runners in ghactions

File: .github/workflows/build-for-pr.yml
Comment (phoinixgrr) [inline .github/workflows/build-for-pr.yml:112]: Maybe decouple the .devDependencies.electron extract with a variable, and use the Variable to inject this to the command. 
Would something like the following work: ? 
```
env:
  DEPENDENCIES: $(jq -r .devDependencies.electron package.json)
```

File: .github/workflows/build-for-pr.yml
Comment (phoinixgrr) [inline .github/workflows/build-for-pr.yml:79]: Maybe decouple the .devDependencies.electron extract with a variable, and use the Variable to inject this to the command. 
Would something like the following work: ? 
```
env:
  DEPENDENCIES: $(jq -r .devDependencies.electron package.json)
```

File: .github/workflows/ci.yaml
Comment (phoinixgrr) [inline .github/workflows/ci.yaml:82]: Maybe decouple the .devDependencies.electron extract with a variable, and use the Variable to inject this to the command. 
Would something like the following work: ? 
```
env:
  DEPENDENCIES: $(jq -r .devDependencies.electron package.json)
```

File: .github/workflows/ci.yaml
Comment (phoinixgrr) [inline .github/workflows/ci.yaml:115]: Maybe decouple the .devDependencies.electron extract with a variable, and use the Variable to inject this to the command. 
Would something like the following work: ? 
```
env:
  DEPENDENCIES: $(jq -r .devDependencies.electron package.json)
```

File: .github/workflows/nightly-browser-view.yml
Comment (phoinixgrr) [inline .github/workflows/nightly-browser-view.yml:171]: Maybe decouple the .version extract with a variable, and use the Variable to inject this to the command. 
Would something like the following work: ? 
```
env:
  APP_VERSION: $(jq -r .version package.json))
```

File: .github/workflows/nightly-browser-view.yml
Comment (phoinixgrr) [inline .github/workflows/nightly-browser-view.yml:213]: Maybe decouple the .version extract with a variable, and use the Variable to inject this to the command. 
Would something like the following work: ? 
```
env:
  APP_VERSION: $(jq -r .version package.json))
```

File: .github/workflows/nightly-rainforest.yml
Comment (phoinixgrr) [inline .github/workflows/nightly-rainforest.yml:91]: Maybe decouple the .version extract with a variable, and use the Variable to inject this to the command. 
Would something like the following work: ? 
```
env:
  APP_VERSION: $(jq -r .version package.json))
```

File: .github/workflows/release.yaml
Comment (phoinixgrr) [inline .github/workflows/release.yaml:148]: Maybe decouple the .version extract with a variable, and use the Variable to inject this to the command. 
Would something like the following work: ? 
```
env:
  APP_VERSION: $(jq -r .version package.json))
```

**Review by phoinixgrr — COMMENTED** [review]
Absolutely AWESOME work here. A huge project. Left some minor comments. 

🎖️ 🎖️ 🎖️ 
Kudos for all the efforts to migrate this beast.

File: .github/actions/patch-nightly-version/action.yaml
Comment (toninis) [reply .github/actions/patch-nightly-version/action.yaml:10]: When you are using composite actions this is mandatory . It can be removed though from the main workflow file

File: .github/actions/patch-nightly-version/action.yaml
Comment (toninis) [reply .github/actions/patch-nightly-version/action.yaml:9]: Nope

File: .github/actions/test/action.yaml
Comment (toninis) [reply .github/actions/test/action.yaml:9]: Nope it defaults to gitbash for Windows

File: .github/workflows/build-for-pr.yml
Comment (toninis) [reply .github/workflows/build-for-pr.yml:35]: Just copied this as is . It might fail sometimes but it's rare . If it fails the next command will fail also if the cache is not updated in any case

File: .github/workflows/build-for-pr.yml
Comment (toninis) [reply .github/workflows/build-for-pr.yml:34]: I agree in general but I suggest we skip this for now and tackle this in the improvement phase, We 've discussed with @devinbinnie about some improvements we need to do in the CI/CD process

File: package.json
Comment (toninis) [reply package.json:53]: I have no input on this . @devinbinnie for input

File: .github/workflows/build-for-pr.yml
Comment (toninis) [reply .github/workflows/build-for-pr.yml:33]: Yep . It works perfectly . I am not so sure if we need to update this

File: .github/workflows/ci.yaml
Comment (toninis) [reply .github/workflows/ci.yaml:115]: You cannot run commands there . But you can do it either in the same step or a previous one

**Review by m1lt0n — APPROVED** [review]

File: .github/actions/patch-nightly-version/action.yaml
Comment (phoinixgrr) [reply .github/actions/patch-nightly-version/action.yaml:9]: Let's remove it if it is unused

File: .github/actions/patch-nightly-version/action.yaml
Comment (phoinixgrr) [reply .github/actions/patch-nightly-version/action.yaml:10]: Got it. Thanks!

**Review by phoinixgrr — APPROVED** [review]
:shipit:

**Pull request merged**
